### ISSUE TITLE
Option to prevent formspecs to be closed

### ISSUE THREAD

**Issue by Zughy** [minetest/minetest#10380]
## Problem

I can't avoid quitting a formspec if I press Esc. I tried calling the same formspec when fields.quit and field.enter_key are received, but if I spam hard enough to trigger the Esc menu, I'm back to zero

## Solutions

An option such as `cant_quit[true]` which is, by default, false

## Additional context

Use-case: avoid players instantly respawning in a minigame

**Comment by v-rob** [comment]
Reshowing in a globalstep might work.  This feature would be simple to add if there's support for it.

**Comment by sorcerykid** [comment]
I'd consider this essential for my Pluggable Helpers management console. The "game" is entirely formspec based, yete right now the only workaround is to kick players that press ESC. It's also needed for my virtual nightclub server since right now new players can skip filling-out their personal profile, even though it's a requirement. There are many use-cases for this feature.

**Comment by SmallJoker** [comment]
But players can currently not return to the server list (ESC menu) when a formspec is shown. Overlaying multiple GUIs must be possible prior implementing this feature.

**Comment by sorcerykid** [comment]
Yes, I was going to suggest that there needs to be a way to force-quit out of the game, preventing rogue mods from locking up the client interface. Perhaps implementing CTRL-ESC would afford a safeguard to immediately return to the server list.

**Comment by Zughy** [comment]
How about having the Esc menu as the only exception to visualise over other formspecs? I can't think of an use-case where you want more than a formspec showed; like, stacked

**Comment by v-rob** [comment]
The Esc menu also uses `GUIFormSpecMenu` since it needs to support prepends, which makes things difficult since it is a monolithic class that acts alone.  Trying to have two `GUIFormSpecMenu`s at once would cause extreme conflict.  Overlaying formspecs is not feasible at all at the present time.  Ctrl+Esc or similar would seem to be the best solution at present.

**Comment by SmallJoker** [comment]
How would this be implemented in Android? They would need a special button to press.

**Comment by sorcerykid** [comment]
I'm with v-rob, and am in favour of keeping this as simple as possible. As for mobile clients, I'd contemplated that possibility as well, but couldn't come up with a solution. I guess the obvious workaround would be to just force close the entire app. Not sure what else could work effectively (and be implemented easily), other than a special button.

**Comment by rubenwardy** [comment]
They could press back twice. This would only work if back wasn't consumed at all (ie: no fields.quit) and you'd probably want to show an Android dialog to confirm

**Comment by v-rob** [comment]
Withdrawing my support.  I have no objections to it, but I think the cons outweigh the pros for now.

**Comment by Zughy** [comment]
I was playing 2021 Jam games and I remembered about this bug: I can break any game that tries to force players to display a formspec by just spamming Esc fast enough, this is ridiculous - and those developers have no fault. Please consider labelling this issue as high priority

**Comment by sfan5** [comment]
I agree that this is a big problem but the fix will need some thought.
I think that the Esc menu should always be reachable, so blocking that is not an option.

> The Esc menu also uses GUIFormSpecMenu since it needs to support prepends, which makes things difficult since it is a monolithic class that acts alone. Trying to have two GUIFormSpecMenus at once would cause extreme conflict. 

What does it do that causes conflicts in global state? I find it hard to believe that this would be *totally* broken when it was initially designed in an object oriented way.

**Comment by LoneWolfHT** [comment]
Instead of showing two menus at once why not reopen menus after the overriding one is closed?

**Comment by v-rob** [comment]
> What does it do that causes conflicts in global state? I find it hard to believe that this would be totally broken when it was initially designed in an object oriented way.

I can't guarantee the conflict exists, but I'm pretty sure `GUIFormSpecMenu` sets itself as the root element in Irrlicht's GUI, and since Irrlicht's GUI is global, all these things `GUIFormSpecMenu` takes for granted (being the root of focus and focus being applied to children only, etc.) would no longer be guarantees.  There's also very possibly other global state issues.

> Instead of showing two menus at once why not reopen menus after the overriding one is closed?

I like this solution; it shouldn't be too difficult to maintain two formspec strings at the same time, one normal and one overriding.

One last thing: if #8679 is merged first, we can just add this as an option to `options[]` instead of adding another entirely new element.

**Comment by Zughy** [comment]
@v-rob is it possible to visualise a fake formspec below when the Esc one is selected? Like, with the engine taking a picture of the previous formspec, keeping it there to give the illusion that is still open, and then remove it as soon as the Esc menu is closed?
Because it'd be weird for players to open a formspec to close another. Say that I go AFK for 5 minutes with the game paused. Looking back at the screen, how am I able to recall that I had a formspec open till I don't actually press Esc to unpause?

**Comment by rollerozxa** [comment]
What's the current consensus on adding this? This is essential for improving the experience with any game that implements formspecs you *do not want to be closed with esc*, whether it be a "game main menu"-type formspec or a formspec simply used for effect (e.g. doing a fade-out effect by taking advantage of formspec fullscreen background). You can of course constantly send the same formspec in a global step, but it will still flicker if you hold it down:

https://github.com/minetest/minetest/assets/60856959/db0b7302-0f09-408c-b452-0025eb3bc6e3

The point of the video is not that you can move the camera around, I could work around that or claim moving the background to be a feature, but that I as a game developer have little to no control over something as simple as wanting to keep a formspec on screen until I myself close it from code.

I obviously do not want to trap players with it, and if a game does that it should be considered a bug in the game, but nothing that simply closing the Minetest window could fix. It's already pretty difficult to time the pause menu when repeatedly showing a formspec on globalstep anyways, so it's just about being able to disable the "hold down for potentially seizure inducing flickering" button.

**Comment by Zughy** [comment]
Also, people can always Alt+F4, terminate the app or (worst case scenario if they're not tech-savvy) reboot the machine. It's game designer's fault if people have no way to quit a formspec when they're expected do, it shouldn't be up to Minetest to take care of the issue

**Comment by appgurueu** [comment]
> Game designers in this case are People who run foreign servers, them beeing able to "trap" players is unacceptable.

They aren't able to trap players. Players will still be able to exit at a window manager level. To recap, here's what that means on common platforms:

* Phone (Android): Home button -> swipe away. The usual workflow to kill an app. I see no issue here. I believe using MT's built in "flow" to exit would be more awkward. These users are who I would be worried about the most, because they are often the least tech-savy. But they are fine.
* "PC" (Windows, Linux): Key combinations like Alt + F4, just hitting the X if not maximized, otherwise again WM key combos.
* Worst case: You go on a "malicious" server, are extremely not-tech-savy, and end up rebooting because you see no other way to fix the problem.

"Unacceptable" doesn't cut it at all. "Poor UX" - and this is just in the hypothetical event of a buggy / malicious server - sure.

> I don't see what the issue is with having them be non-closeable, in the sense that if you press escape the main menu pops up, and if you press it again you switch back. There is no need to show the old formspec in the background or anything, just make sure the main menu is _always_ accesible

1. I'm not convinced "ESC to swap formspec & mainmenu" can be considered good UX.
2. As is always the case with Minetest, things are usually much less "trivial" than they seem. If you believe that this is "trivial" to do properly, please demonstrate it; my gut feeling is that it takes more effort than I'm willing to make for this issue.

**Comment by v-rob** [comment]
I just made a simple implementation of this. The mechanisms (`m_allowclose`) are already in place for the main menu, so it's just a matter of exposing it to the user and adding `uncloseable[true]` to every main menu formspec rather than having it hardcoded.

The remaining question is how to handle the pause menu. I don't think that pressing escape on an uncloseable formspec should open the pause menu, because the modder may want to give special handling to `fields.quit` (for example, to show a confirmation dialog or have something else happen without the formspec closing).

There is the option of having ctrl + escape open the pause menu from any formspec (not just uncloseable ones), but many people are likely to be unaware of its existence. Alternatively, the pause menu is rendered inaccessible, which is suboptimal (disregarding malicious servers, you might want to exit the game in the middle of a long modal dialog).

I'm happy to make this a PR, but this needs addressing.

**Comment by Wuzzy2** [comment]
If you add this feature, there definitely should be an "emergency exit" option so players can escape from a buggy game.

I think the biggest flaw is that the Esc menu is itself just a formspec and it can be replaced by a simple minetest.show_formspec in multiplayer. This is wrong (bug was reported elsewhere, but I think this is part of a bigger design flaw).

The pause menu should have a special treatment in Luanti, it should essentially be "on top" of the in-game formspecs; i.e. the Lua API simply cannot interact with the pause menu at all, and vice-versa. I mean, if the pause menu is open, everything of the game, *including its formspecs* go to the background (at least in a logical/input sense; it would be a OK solution to hide the game formspecs while the pause menu is open).

Once the pause menu has become fully independent of the game formspecs, I think the Ctrl+Escape emergency exit would be a first good option.

More daring is the idea of changing the key to exit an in-game formspec. Maybe one of the function keys? Tab? And then Esc is reserved *exclusively* for the pause menu (so an Esc press is never "seen" by the game, only by Luanti), and it can be opened even if a formspec is open. I am sure this idea will be very controversial.

**Comment by LoneWolfHT** [comment]
> because the modder may want to give special handling to fields.quit
I'd use a normal formspec if I wanted that.

I think it's reasonable to reserve the pause menu for the ESC key/mobile back button in unclosable formspecs.

**Comment by v-rob** [comment]
> > because the modder may want to give special handling to fields.quit
> 
> I'd use a normal formspec if I wanted that.

Then you're back to square one with flickery land where you are able to interact with the camera/world between formspecs.

> Can unclosable formspecs be closed if another formspec (maybe CSM) is shown on top of it?

Yes by design (otherwise how do you close/update the formspec?), but how mods choose to detect conflicting mods closing their formspecs is a problem more along the lines of #13142.

**Comment by grorp** [comment]
>  I tried calling the same formspec when fields.quit and field.enter_key are received, but if I spam hard enough to trigger the Esc menu, I'm back to zero

For reference: That was #11907 btw
